Add new -concat_nested_namespaces option

vmiklos opened the pull request:
This implements the idea from
<https://groups.google.com/forum/#!topic/include-what-you-use/-jg0QwmxdK8>.

Is the idea OK in general? If so, happy to spend time on an actual testcase.

vmiklos commented:
Short demo on how this works:

````
$ ~/git/include-what-you-use/workdir/bin/include-what-you-use -Xiwyu --concat_nested_namespaces -std=c++17 two.cpp

two.h should add these lines:
namespace a { struct One3; }
namespace a::b { struct One2; }
namespace a::b::c { struct One; }

two.h should remove these lines:
- #include "one.h"  // lines 1-1

The full include-list for two.h:
namespace a { struct One3; }
namespace a::b { struct One2; }
namespace a::b::c { struct One; }
---

two.cpp should add these lines:
namespace a { struct One3; }
namespace a::b { struct One2; }
namespace a::b::c { struct One; }

two.cpp should remove these lines:

The full include-list for two.cpp:
#include "two.h"
namespace a { struct One3; }
namespace a::b { struct One2; }
namespace a::b::c { struct One; }
---
$ ~/git/include-what-you-use/workdir/bin/include-what-you-use -std=c++17 two.cpp

two.h should add these lines:
namespace a { namespace b { namespace c { struct One; } } }
namespace a { namespace b { struct One2; } }
namespace a { struct One3; }

two.h should remove these lines:
- #include "one.h"  // lines 1-1

The full include-list for two.h:
namespace a { namespace b { namespace c { struct One; } } }
namespace a { namespace b { struct One2; } }
namespace a { struct One3; }
---

two.cpp should add these lines:
namespace a { namespace b { namespace c { struct One; } } }
namespace a { namespace b { struct One2; } }
namespace a { struct One3; }

two.cpp should remove these lines:

The full include-list for two.cpp:
#include "two.h"
namespace a { namespace b { namespace c { struct One; } } }
namespace a { namespace b { struct One2; } }
namespace a { struct One3; }
---

````

kimgr commented:
I like this in principle! I was about to suggest to always do this for `-std=c++17`, but that will break portable code bases. But let's find a shorter name for the switch... How about `--cxx17ns`? It's a little arcane, but it's short and easier to type.

Also, the code for namespace formatting is pretty raw, maybe that should be extracted to `iwyu_ast_util`? In fact, maybe the whole `PrintForwardDeclare` should move, and you can extract the namespace formatting more effectively in the AST util module? Something like:

```
std::string PrintForwardDeclare(const NamedDecl*decl,
                                const std::string& tpl_parmas_and_kind,
                                bool cxx17ns)
```

Another, opposing, thought I had when I first saw this: I think IWYU should generally do less of source cleanup (no include sorting, no clang-formatting, no fix_includes.py, etc.) and focus more on just producing a report that can be easily consumed by downstream tools (e.g. `clang-apply-replacements`). We could save so much time and focus our efforts on the IWYU problem if we moved in that direction.

But I think that's still a utopia on the horizon and this patch is small and focused, so let's do it for now. It would be cool, though, if you could create a `clang-tidy` check that did the same transform outside of the realm of IWYU (that way we could eventually retire this.)

vmiklos commented:
> I like this in principle! I was about to suggest to always do this for `-std=c++17`, but that will break portable code bases. But let's find a shorter name for the switch... How about `--cxx17ns`? It's a little arcane, but it's short and easier to type.

Renamed, I'm not good at naming. :-)

> Also, the code for namespace formatting is pretty raw, maybe that should be extracted to `iwyu_ast_util`? In fact, maybe the whole `PrintForwardDeclare` should move, and you can extract the namespace formatting more effectively in the AST util module? Something like:
> 
> ```
> std::string PrintForwardDeclare(const NamedDecl*decl,
>                                 const std::string& tpl_params_and_kind,
>                                 bool cxx17ns)
> ```

I can do that, but PrintForwardDeclare() depends on FakeNamedDecl, which is still in iwyu_output, should I move that as well? Or let iwyu_ast_util depend on iwyu_output?

> Another, opposing, thought I had when I first saw this: I think IWYU should generally do less of source cleanup (no include sorting, no clang-formatting, no fix_includes.py, etc.) and focus more on just producing a report that can be easily consumed by downstream tools (e.g. `clang-apply-replacements`). We could save so much time and focus our efforts on the IWYU problem if we moved in that direction.
> 
> But I think that's still a utopia on the horizon and this patch is small and focused, so let's do it for now. It would be cool, though, if you could create a `clang-tidy` check that did the same transform outside of the realm of IWYU (that way we could eventually retire this.)

clang-tidy already has a modernize-concat-nested-namespaces check. The problem comes once that clang-tidy check was already ran on a codebase, and IWYU suggests new forward-declarations in the old style, which feels inconsistent.

Perhaps in the long run the clang AST API could give us this thing (either in the old or in the new C++17 form), after all it already supports giving us the fully qualified name, which is a bit similar.

(Testcase is still not yet added, but I plan to look at that.)

kimgr commented:
> clang-tidy already has a modernize-concat-nested-namespaces check.

Oh, cool.

> The problem comes once that clang-tidy check was already ran on a
> codebase, and IWYU suggests new forward-declarations in the old style,
> which feels inconsistent.

Yeah, I meant that it would be better if we could chain these tools so that clang-tidy could run after IWYU and fix the decls back up.

It would be nice if IWYU could focus on the core problem of IWYU, and not worry so much about rewriting.

vmiklos commented on iwyu_output.cc:
```diff
@@ -367,14 +368,33 @@ string PrintForwardDeclare(const NamedDecl* decl,
+          first = false;
+          prefix = prefix + " { ";
+        }
+        if (ctx->getParent() && isa<NamedDecl>(ctx->getParent())) {
```
Agreed, I don't think it can be anything else in practice, but NamedDecl is consistent with the above check in the for loop, so sounds like a good idea to keep it as-is.

vmiklos commented on iwyu_output.cc:
```diff
@@ -367,14 +368,33 @@ string PrintForwardDeclare(const NamedDecl* decl,
+        fwd_decl = prefix + fwd_decl + suffix;
+      } else {
+        const std::string ns_name = ns->isAnonymousNamespace() ?
+            "" : (std::string(ns->getName()) + " ");
```
Done.

kimgr commented:
Thanks, this looks good. Let's merge once the testcase is in place.

vmiklos commented:
> Thanks, this looks good. Let's merge once the testcase is in place.

I've added that now.

kimgr commented on tests/cxx/cxx17ns.cc:
```diff
@@ -0,0 +1,33 @@
+/**** IWYU_SUMMARY
+
+tests/cxx/cxx17ns.cc should add these lines:
+namespace a { struct One3; }
```
Nice testcase! Is it possible to add coverage for the anonymous namespace cases too? I don't know if this would trigger that path:

    namespace { class Unused; }

but

    namespace a {
    namespace {
        struct Three {};
    }
    ...
    }

might.

vmiklos replied:
Oh, this is a very good point. So https://clang.llvm.org/extra/clang-tidy/checks/modernize-concat-nested-namespaces.html explicitly doesn't use that syntax for anonymous namespaces; I'll avoid the new syntax for that case as well.

vmiklos commented:
Ah, and one more thing, I should avoid "namespace  {" (double spaces). Will fix in a moment.

kimgr commented:
Thanks, this diff looks good to me!

The commit message could use a touch-up, let's explain the problem and solution there instead of referring to an outside resource.

It just occurred to me that `fix_includes.py` also does some munging of namespaces, so I wonder if it needs to be adjusted to match this. Sorry I didn't think about this earlier. I looked it over quickly, and it actually looks like it might work, could you test it? Basically `include-what-you-use scratchfile.cpp 2>&1 | fix_includes.py` (will rewrite scratchfile.cpp in place, so be careful).

Thanks!

vmiklos commented:
````
workdir/bin/include-what-you-use -Xiwyu --cxx17ns -std=c++17 -I. tests/cxx/cxx17ns.cc 2>&1 | python fix_includes.py
````

Is what I tested, the result looks good to me.

kimgr commented:
Cool! Maybe it would be prudent to add a test case to fix_includes_test.py to preserve this behavior.

vmiklos commented:
> Maybe it would be prudent to add a test case to fix_includes_test.py to preserve this behavior.

Done now.

kimgr commented:
Beautiful, thank you!
